<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Frameset//EN""http://www.w3.org/TR/REC-html40/frameset.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc on Sat Mar 09 17:00:21 CST 2002 -->
<TITLE>
FORESTER 1.92 API specification: Class  RIO
</TITLE>
<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">
</HEAD>
<BODY BGCOLOR="white">

<!-- ========== START OF NAVBAR ========== -->
<A NAME="navbar_top"><!-- --></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
FORESTER 1.92</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../forester/tools/OE.html"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../forester/tools/SDI.html"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" TARGET="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RIO.html" TARGET="_top"><B>NO FRAMES</B></A></FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY: &nbsp;INNER&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL: &nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<!-- =========== END OF NAVBAR =========== -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
forester.tools</FONT>
<BR>
Class  RIO</H2>
<PRE>
java.lang.Object
  |
  +--<B>forester.tools.RIO</B>
</PRE>
<HR>
<DL>
<DT>public class <B>RIO</B><DT>extends java.lang.Object</DL>

<P>
<DL>
<DT><B>Version: </B><DD>1.400 -- last modified: 12/27/01</DD>
<DT><B>Author: </B><DD>Christian M. Zmasek</DD>
</DL>
<HR>

<P>
<!-- ======== INNER CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->


<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#RIO()">RIO</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Default constructor.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getBootstraps()">getBootstraps</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the numbers of trees analyzed.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getDistance(java.lang.String)">getDistance</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the distance to a sequences/taxa after
a distance list file has been read in with readDistanceList(File).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getDistance(java.lang.String, java.lang.String)">getDistance</A></B>(java.lang.String&nbsp;name1,
            java.lang.String&nbsp;name2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the distance between two sequences/taxa after
a distance matrix file has been read in with readDistanceMatrix(File).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getExtNodesOfAnalyzedGeneTrees()">getExtNodesOfAnalyzedGeneTrees</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the numbers of number of ext
nodes in gene trees analyzed (after stripping).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getInferredOrthologs(java.lang.String)">getInferredOrthologs</A></B>(java.lang.String&nbsp;seq_name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a HashMap containing the inferred orthologs of the external gene
tree node with the sequence name seq_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getInferredSubtreeNeighbors(java.lang.String)">getInferredSubtreeNeighbors</A></B>(java.lang.String&nbsp;seq_name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a HashMap containing the inferred "subtree neighbors" of the external 
gene tree node with the sequence name seq_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getInferredSuperOrthologs(java.lang.String)">getInferredSuperOrthologs</A></B>(java.lang.String&nbsp;seq_name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a HashMap containing the inferred "super orthologs" of the external 
gene tree node with the sequence name seq_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getInferredUltraParalogs(java.lang.String)">getInferredUltraParalogs</A></B>(java.lang.String&nbsp;seq_name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a HashMap containing the inferred "ultra paralogs" of the external 
gene tree node with the sequence name seq_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getOrder(int)">getOrder</A></B>(int&nbsp;sort)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the order in which ortholog (o), 
"super ortholog" (s) and distance (d) are
returned and sorted (priority of sort always
goes from left to right), given sort.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#getTime()">getTime</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the time (in ms) needed to run "inferOrthologs".</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)">inferOrthologs</A></B>(java.io.File&nbsp;gene_trees_file,
               <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Infers the orthologs (as well the "super orthologs", the "x orthologs",
and the "ultra paralogs") for each external node in Tree array gene_trees.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree, java.lang.String)">inferOrthologs</A></B>(java.io.File&nbsp;gene_trees_file,
               <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree,
               java.lang.String&nbsp;query)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Infers the orthologs (as well the "super orthologs", the "subtree neighbors",
and the "ultra paralogs") for each external node of the gene Trees 
in multiple tree File gene_trees_file
(=output of PHYLIP NEIGHBOR, for example).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)">inferOrthologs</A></B>(<A HREF="../../forester/tree/Tree.html">Tree</A>[]&nbsp;gene_trees,
               <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Infers the orthologs (as well the "super orthologs", the "x orthologs",
and the "ultra paralogs") for each external node in Tree array gene_trees.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.ArrayList</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferredOrthologsToArrayList(java.lang.String, double)">inferredOrthologsToArrayList</A></B>(java.lang.String&nbsp;seq_name,
                             double&nbsp;threshold_orthologs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an ArrayList containg the names of orthologs of the Node
with seq name seq_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)">inferredOrthologsToString</A></B>(java.lang.String&nbsp;query_name,
                          int&nbsp;sort,
                          double&nbsp;threshold_orthologs,
                          double&nbsp;threshold_subtreeneighborings)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a String containg the names of orthologs of the Node with seq name
query_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferredOrthologTableToFile(java.io.File)">inferredOrthologTableToFile</A></B>(java.io.File&nbsp;outfile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the orthologs for each external node of the gene trees to
outfile in the form of a table.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferredSuperOrthologTableToFile(java.io.File)">inferredSuperOrthologTableToFile</A></B>(java.io.File&nbsp;outfile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the "super orthologs" for each external nodes of the gene trees to
outfile in the form of a table.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#inferredUltraParalogsToString(java.lang.String, boolean, double)">inferredUltraParalogsToString</A></B>(java.lang.String&nbsp;query_name,
                              boolean&nbsp;return_dists,
                              double&nbsp;threshold_ultra_paralogs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a String containg the names of orthologs of the Node with seq name
query_name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#readDistanceList(java.io.File)">readDistanceList</A></B>(java.io.File&nbsp;dist_list)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in the distance list file dist_list.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#readDistanceMatrix(java.io.File)">readDistanceMatrix</A></B>(java.io.File&nbsp;matrix_file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in the distance matrix file matrix_file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#reset()">reset</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Brings this into the same state as immediately after construction.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../forester/tools/RIO.html#roundToInt(double)">roundToInt</A></B>(double&nbsp;d)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Rounds a double to a int.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><clinit>, clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->


<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="RIO()"><!-- --></A><H3>
RIO</H3>
<PRE>
public <B>RIO</B>()</PRE>
<DL>
<DD>Default constructor.</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="reset()"><!-- --></A><H3>
reset</H3>
<PRE>
public void <B>reset</B>()</PRE>
<DL>
<DD>Brings this into the same state as immediately after construction.</DL>
<HR>

<A NAME="inferOrthologs(java.io.File, forester.tree.Tree, java.lang.String)"><!-- --></A><H3>
inferOrthologs</H3>
<PRE>
public void <B>inferOrthologs</B>(java.io.File&nbsp;gene_trees_file,
                           <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree,
                           java.lang.String&nbsp;query)
                    throws java.lang.Exception</PRE>
<DL>
<DD>Infers the orthologs (as well the "super orthologs", the "subtree neighbors",
and the "ultra paralogs") for each external node of the gene Trees 
in multiple tree File gene_trees_file
(=output of PHYLIP NEIGHBOR, for example).
Tallies how many times each sequence is (super-) orthologous towards
the query. 
Tallies how many times each sequence is ultra paralogous towards
the query. 
Tallies how many times each sequence is a subtree neighbor of
the query.
Gene duplications are inferred using SDI. 
Modifies its argument species_tree.
Is a little faster than "inferOrthologs(File,Tree)" since orthologs
are only inferred for query.
<p>
To obtain the results use the methods listed below. 
<p>
(Last modified: 10/19/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gene_trees_file</CODE> - a File containing gene Trees in NH format, which
is the result of performing a bootstrap analysis
in PHYLIP<DD><CODE>species_tree</CODE> - a species Tree, which has species names in
its species fields<DD><CODE>query</CODE> - the sequence name of the squence whose orthologs
are to be inferred<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#getInferredOrthologs(java.lang.String)"><CODE>getInferredOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getInferredSuperOrthologs(java.lang.String)"><CODE>getInferredSuperOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)"><CODE>inferredOrthologsToString(String,int,double,double)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologTableToFile(java.io.File)"><CODE>inferredOrthologTableToFile(File)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredSuperOrthologTableToFile(java.io.File)"><CODE>inferredSuperOrthologTableToFile(File)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><!-- --></A><H3>
inferOrthologs</H3>
<PRE>
public void <B>inferOrthologs</B>(<A HREF="../../forester/tree/Tree.html">Tree</A>[]&nbsp;gene_trees,
                           <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree)
                    throws java.lang.Exception</PRE>
<DL>
<DD>Infers the orthologs (as well the "super orthologs", the "x orthologs",
and the "ultra paralogs") for each external node in Tree array gene_trees.
Tallies how many times each sequence is (super-) orthologous towards
all other ones. 
Tallies how many times each sequence is super paralogous towards
all other ones.
Tallies how many times each sequence is a subtree neighbor of
the query.
Gene duplications are inferred using SDI. 
Modifies its arguments gene_trees and species_tree.
Method "inferOrthologs(File,Tree)" accomplishes the same, but 
needs less memory since not all gene Trees have to be kept in memory.
<p>
To obtain the results use the methods listed below. 
<p>
(Last modified: 12/07/00)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gene_trees</CODE> - a Tree[] which is the result of performing 
a bootstrap analysis, nodes of trees must have
species names and sequence names in the
appropriate fields<DD><CODE>species_tree</CODE> - a species Tree, which has species names in
its species fields<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#getInferredOrthologs(java.lang.String)"><CODE>getInferredOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getInferredSuperOrthologs(java.lang.String)"><CODE>getInferredSuperOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)"><CODE>inferredOrthologsToString(String,int,double,double)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologTableToFile(java.io.File)"><CODE>inferredOrthologTableToFile(File)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredSuperOrthologTableToFile(java.io.File)"><CODE>inferredSuperOrthologTableToFile(File)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferOrthologs(java.io.File, forester.tree.Tree)"><!-- --></A><H3>
inferOrthologs</H3>
<PRE>
public void <B>inferOrthologs</B>(java.io.File&nbsp;gene_trees_file,
                           <A HREF="../../forester/tree/Tree.html">Tree</A>&nbsp;species_tree)
                    throws java.lang.Exception</PRE>
<DL>
<DD>Infers the orthologs (as well the "super orthologs", the "x orthologs",
and the "ultra paralogs") for each external node in Tree array gene_trees.
Tallies how many times each sequence is (super-) orthologous towards
all other ones. 
Tallies how many times each sequence is ultra paralogous towards
all other ones.
Tallies how many times each sequence is a subtree neighbor of
the query.
Gene duplications are inferred using SDI. 
Modifies its argument species_tree.
Might be a little slower than "inferOrthologs(File,Tree,String)".
Needs less memory than "inferOrthologs(Tree[],Tree)" since not all
gene Trees have to be kept in memory.
<p>
To obtain the results use the methods listed below. 
<p>
(Last modified: 10/19/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gene_trees_file</CODE> - a File containing gene Trees in NH format, which
is the result of performing a bootstrap analysis
in PHYLIP<DD><CODE>species_tree</CODE> - a species Tree, which has species names in
its species fields<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree, java.lang.String)"><CODE>inferOrthologs(File,Tree,String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getInferredOrthologs(java.lang.String)"><CODE>getInferredOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getInferredSuperOrthologs(java.lang.String)"><CODE>getInferredSuperOrthologs(String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)"><CODE>inferredOrthologsToString(String,int,double,double)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologTableToFile(java.io.File)"><CODE>inferredOrthologTableToFile(File)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredSuperOrthologTableToFile(java.io.File)"><CODE>inferredSuperOrthologTableToFile(File)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferredOrthologsToString(java.lang.String, int, double, double)"><!-- --></A><H3>
inferredOrthologsToString</H3>
<PRE>
public java.lang.String <B>inferredOrthologsToString</B>(java.lang.String&nbsp;query_name,
                                                  int&nbsp;sort,
                                                  double&nbsp;threshold_orthologs,
                                                  double&nbsp;threshold_subtreeneighborings)
                                           throws java.lang.Exception</PRE>
<DL>
<DD>Returns a String containg the names of orthologs of the Node with seq name
query_name. The String also contains how many times a particular ortholog has
been observed.
<p>
<ul> 
The output order is (per line):
Name, Ortholog, Subtree neighbor, Super ortholog, Distance
</ul>
<p>
The sort priority of this is determined by sort
in the following manner:
<ul>

<li> 0 : Ortholog  
<li> 1 : Ortholog,         Super ortholog
<li> 2 : Super ortholog,   Ortholog
<li> 3 : Ortholog,         Distance
<li> 4 : Distance,         Ortholog
<li> 5 : Ortholog,         Super ortholog,   Distance  
<li> 6 : Ortholog,         Distance,         Super ortholog
<li> 7 : Super ortholog,   Ortholog,         Distance
<li> 8 : Super ortholog,   Distance,         Ortholog
<li> 9 : Distance,         Ortholog,         Super ortholog
<li>10 : Distance,         Super ortholog,   Ortholog
<li>11 : Ortholog,         Subtree neighbor, Distance 
<li>12 : Ortholog,         Subtree neighbor, Super ortholog,   Distance (default)
<li>13 : Ortholog,         Super ortholog,   Subtree neighbor, Distance
<li>14 : Subtree neighbor, Ortholog,         Super ortholog,   Distance
<li>15 : Subtree neighbor, Distance,         Ortholog,         Super ortholog   
<li>16 : Ortholog,         Distance,         Subtree neighbor, Super ortholog 
<li>17 : Ortholog,         Subtree neighbor, Distance,         Super ortholog 

</ul>
<p>
Returns "-" if no putative orthologs have been found
(given threshold_orthologs).
<p>
Orthologs are to be inferred by method "inferOrthologs".
<p>
(Last modified: 05/08/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>query_name</CODE> - sequence name of a external node of the gene trees<DD><CODE>sort</CODE> - order and sort priority<DD><CODE>threshold_orthologs</CODE> - the minimal number of observations for a
a sequence to be reported as orthologous,
in percents (0.0-100.0%)<DD><CODE>threshold_subtreeneighborings</CODE> - the minimal number of observations for a
a sequence to be reported as orthologous,
in percents (0.0-100.0%)<DT><B>Returns:</B><DD>String containing the inferred orthologs,
String containing "-" if no orthologs have been found
null in case of error<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree, java.lang.String)"><CODE>inferOrthologs(File,Tree,String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getOrder(int)"><CODE>getOrder(int)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferredUltraParalogsToString(java.lang.String, boolean, double)"><!-- --></A><H3>
inferredUltraParalogsToString</H3>
<PRE>
public java.lang.String <B>inferredUltraParalogsToString</B>(java.lang.String&nbsp;query_name,
                                                      boolean&nbsp;return_dists,
                                                      double&nbsp;threshold_ultra_paralogs)
                                               throws java.lang.Exception</PRE>
<DL>
<DD>Returns a String containg the names of orthologs of the Node with seq name
query_name. The String also contains how many times a particular ortholog has
been observed. 

Returns "-" if no putative orthologs have been found
(given threshold_orthologs).
<p>
Orthologs are to be inferred by method "inferOrthologs".
<p>
(Last modified: 10/16/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>query_name</CODE> - sequence name of a external node of the gene trees<DD><CODE>return_dists</CODE> - &nbsp;<DD><CODE>threshold_ultra_paralogs</CODE> - between 1 and 100<DT><B>Returns:</B><DD>String containing the inferred orthologs,
String containing "-" if no orthologs have been found
null in case of error<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree, java.lang.String)"><CODE>inferOrthologs(File,Tree,String)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getOrder(int)"><CODE>getOrder(int)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferredOrthologsToArrayList(java.lang.String, double)"><!-- --></A><H3>
inferredOrthologsToArrayList</H3>
<PRE>
public java.util.ArrayList <B>inferredOrthologsToArrayList</B>(java.lang.String&nbsp;seq_name,
                                                        double&nbsp;threshold_orthologs)
                                                 throws java.lang.Exception</PRE>
<DL>
<DD>Returns an ArrayList containg the names of orthologs of the Node
with seq name seq_name.
<p>
(Last modified: 10/07/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seq_name</CODE> - sequence name of a external node of the gene trees<DD><CODE>threshold_orthologs</CODE> - the minimal number of observations for a
a sequence to be reported as orthologous
as percentage (0.0-100.0%)<DT><B>Returns:</B><DD>ArrayList containg the names of orthologs of the Node with seq name
seq_name<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#getOrder(int)"><CODE>getOrder(int)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferredOrthologTableToFile(java.io.File)"><!-- --></A><H3>
inferredOrthologTableToFile</H3>
<PRE>
public void <B>inferredOrthologTableToFile</B>(java.io.File&nbsp;outfile)
                                 throws java.lang.Exception</PRE>
<DL>
<DD>Writes the orthologs for each external node of the gene trees to
outfile in the form of a table.
Orthologs are to be inferred by method "inferOrthologs".
Overwrites without asking!
(Last modified: 12/07/00)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>outfile</CODE> - the File to write to<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="inferredSuperOrthologTableToFile(java.io.File)"><!-- --></A><H3>
inferredSuperOrthologTableToFile</H3>
<PRE>
public void <B>inferredSuperOrthologTableToFile</B>(java.io.File&nbsp;outfile)
                                      throws java.lang.Exception</PRE>
<DL>
<DD>Writes the "super orthologs" for each external nodes of the gene trees to
outfile in the form of a table.
Super orthologs are to be inferred by method "inferOrthologs".
Overwrites without asking!
(Last modified: 11/28/00)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>outfile</CODE> - the File to write to<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getInferredOrthologs(java.lang.String)"><!-- --></A><H3>
getInferredOrthologs</H3>
<PRE>
public java.util.HashMap <B>getInferredOrthologs</B>(java.lang.String&nbsp;seq_name)
                                       throws java.lang.Exception</PRE>
<DL>
<DD>Returns a HashMap containing the inferred orthologs of the external gene
tree node with the sequence name seq_name. Sequence names
are the keys (String), numbers of observations are the values (Int).
Orthologs are to be inferred by method "inferOrthologs".
Throws an exception if seq_name is not found.
(Last modified: 05/08/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seq_name</CODE> - sequence name of a external node of the gene trees<DT><B>Returns:</B><DD>HashMap containing the inferred orthologs (name(String)->value(Int))<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getInferredSuperOrthologs(java.lang.String)"><!-- --></A><H3>
getInferredSuperOrthologs</H3>
<PRE>
public java.util.HashMap <B>getInferredSuperOrthologs</B>(java.lang.String&nbsp;seq_name)
                                            throws java.lang.Exception</PRE>
<DL>
<DD>Returns a HashMap containing the inferred "super orthologs" of the external 
gene tree node with the sequence name seq_name. Sequence names
are the keys (String), numbers of observations are the values (Int).
Super orthologs are to be inferred by method "inferOrthologs".
Throws an exception if seq_name is not found.
(Last modified: 05/08/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seq_name</CODE> - sequence name of a external node of the gene trees<DT><B>Returns:</B><DD>HashMap containing the inferred super orthologs
(name(String)->value(Int))<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getInferredSubtreeNeighbors(java.lang.String)"><!-- --></A><H3>
getInferredSubtreeNeighbors</H3>
<PRE>
public java.util.HashMap <B>getInferredSubtreeNeighbors</B>(java.lang.String&nbsp;seq_name)
                                              throws java.lang.Exception</PRE>
<DL>
<DD>Returns a HashMap containing the inferred "subtree neighbors" of the external 
gene tree node with the sequence name seq_name. Sequence names
are the keys (String), numbers of observations are the values (Int).
"subtree neighbors" are to be inferred by method "inferOrthologs".
Throws an exception if seq_name is not found.
(Last modified: 12/26/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seq_name</CODE> - sequence name of a external node of the gene trees<DT><B>Returns:</B><DD>HashMap containing the inferred "subtree neighbors"
(name(String)->value(Int))<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getInferredUltraParalogs(java.lang.String)"><!-- --></A><H3>
getInferredUltraParalogs</H3>
<PRE>
public java.util.HashMap <B>getInferredUltraParalogs</B>(java.lang.String&nbsp;seq_name)
                                           throws java.lang.Exception</PRE>
<DL>
<DD>Returns a HashMap containing the inferred "ultra paralogs" of the external 
gene tree node with the sequence name seq_name. Sequence names
are the keys (String), numbers of observations are the values (Int).
"ultra paralogs" are to be inferred by method "inferOrthologs".
Throws an exception if seq_name is not found.
(Last modified: 05/08/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seq_name</CODE> - sequence name of a external node of the gene trees<DT><B>Returns:</B><DD>HashMap containing the inferred ultra paralogs
(name(String)->value(Int))<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getOrder(int)"><!-- --></A><H3>
getOrder</H3>
<PRE>
public java.lang.String <B>getOrder</B>(int&nbsp;sort)</PRE>
<DL>
<DD>Returns the order in which ortholog (o), 
"super ortholog" (s) and distance (d) are
returned and sorted (priority of sort always
goes from left to right), given sort.
For the meaning of sort<DD><DL>
<DT><B>Parameters:</B><DD><CODE>sort</CODE> - determines order and sort priority<DT><B>Returns:</B><DD>String indicating the order<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)"><CODE><p>
(Last modified: 10/07/01)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferredOrthologsToString(java.lang.String, int, double, double)"><CODE>inferredOrthologsToString(String,int,double,double)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getTime()"><!-- --></A><H3>
getTime</H3>
<PRE>
public long <B>getTime</B>()</PRE>
<DL>
<DD>Returns the time (in ms) needed to run "inferOrthologs".
Final variable TIME needs to be set to true.
<p>
(Last modified: 11/14/00)<DD><DL>
<DT><B>Returns:</B><DD>time (in ms) needed to run method "inferOrthologs"<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getBootstraps()"><!-- --></A><H3>
getBootstraps</H3>
<PRE>
public int <B>getBootstraps</B>()</PRE>
<DL>
<DD>Returns the numbers of trees analyzed.
<p>
(Last modified: 02/20/00)<DD><DL>
<DT><B>Returns:</B><DD>the numbers of trees analyzed<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getExtNodesOfAnalyzedGeneTrees()"><!-- --></A><H3>
getExtNodesOfAnalyzedGeneTrees</H3>
<PRE>
public int <B>getExtNodesOfAnalyzedGeneTrees</B>()</PRE>
<DL>
<DD>Returns the numbers of number of ext
nodes in gene trees analyzed (after stripping).
<p>
(Last modified: 05/02/00)<DD><DL>
<DT><B>Returns:</B><DD>number of ext nodes in gene trees analyzed (after stripping)<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#inferOrthologs(forester.tree.Tree[], forester.tree.Tree)"><CODE>inferOrthologs(Tree[],Tree)</CODE></A>, 
<A HREF="../../forester/tools/RIO.html#inferOrthologs(java.io.File, forester.tree.Tree)"><CODE>inferOrthologs(File,Tree)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="readDistanceMatrix(java.io.File)"><!-- --></A><H3>
readDistanceMatrix</H3>
<PRE>
public void <B>readDistanceMatrix</B>(java.io.File&nbsp;matrix_file)
                        throws java.lang.Exception</PRE>
<DL>
<DD>Reads in the distance matrix file matrix_file.
The file must have the following format.
<p>
<pre>
[number of sequences, is ignored, and is not necessary]
name x x x 
name x x x
name x x x
</pre>
<p> 
Number of white spaces does not matter.
Such files are produces by PHYLIP PROTDIST. Files 
produced by TREE-PUZZLE need to be modified to remove
the newlines "inside" the rows.
<p>
(Last modified: 11/28/00)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>matrix_file</CODE> - a File containing a distance matrix<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#getDistance(java.lang.String, java.lang.String)"><CODE>getDistance(String,String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="readDistanceList(java.io.File)"><!-- --></A><H3>
readDistanceList</H3>
<PRE>
public void <B>readDistanceList</B>(java.io.File&nbsp;dist_list)
                      throws java.lang.Exception</PRE>
<DL>
<DD>Reads in the distance list file dist_list.
The file must have the following format.
<p>
<pre>
name x
name x
name x
</pre> 
Number of white spaces does not matter.
Such files are produces by modified TREE-PUZZLE.
<p>
(Last modified: 05/30/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>dist_list</CODE> - a File containing a distance matrix<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#getDistance(java.lang.String)"><CODE>getDistance(String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getDistance(java.lang.String, java.lang.String)"><!-- --></A><H3>
getDistance</H3>
<PRE>
public double <B>getDistance</B>(java.lang.String&nbsp;name1,
                          java.lang.String&nbsp;name2)
                   throws java.lang.Exception</PRE>
<DL>
<DD>Returns the distance between two sequences/taxa after
a distance matrix file has been read in with readDistanceMatrix(File).
Throws an exception if name1 or name2 are not found or if
no matrix has been read in.
<p>
(Last modified: 11/28/00)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>name1</CODE> - a sequence name
name2 a sequence name<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#readDistanceMatrix(java.io.File)"><CODE>readDistanceMatrix(File)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getDistance(java.lang.String)"><!-- --></A><H3>
getDistance</H3>
<PRE>
public double <B>getDistance</B>(java.lang.String&nbsp;name)
                   throws java.lang.Exception</PRE>
<DL>
<DD>Returns the distance to a sequences/taxa after
a distance list file has been read in with readDistanceList(File).
Throws an exception if name is not found or if
no list has been read in.
<p>
(Last modified: 05/30/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>name</CODE> - a sequence name<DT><B>See Also: </B><DD><A HREF="../../forester/tools/RIO.html#readDistanceList(java.io.File)"><CODE>readDistanceList(File)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="roundToInt(double)"><!-- --></A><H3>
roundToInt</H3>
<PRE>
public static int <B>roundToInt</B>(double&nbsp;d)</PRE>
<DL>
<DD>Rounds a double to a int.
<p>
(Last modified: 03/15/01)<DD><DL>
<DT><B>Parameters:</B><DD><CODE>d</CODE> - a double to be rounded to a int<DT><B>Returns:</B><DD>d rounded to a int</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>

<!-- ========== START OF NAVBAR ========== -->
<A NAME="navbar_bottom"><!-- --></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
FORESTER 1.92</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../forester/tools/OE.html"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../forester/tools/SDI.html"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" TARGET="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RIO.html" TARGET="_top"><B>NO FRAMES</B></A></FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY: &nbsp;INNER&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL: &nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<!-- =========== END OF NAVBAR =========== -->

<HR>

</BODY>
</HTML>
